### PR TITLE
Pytorch embedding

### PR DESCRIPTION
**Pull request by DeNeutoy**
Switching the word embedding functionality over to use Pytorch. Mainly did this to get a quick idea of how tough it would be to switch layers. Answer: pretty straightforward.

### CODE REVIEW

File: allennlp/layers/embeddings.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+class Embedding(torch.nn.Module):
+    """
+    A more featureful embedding module than the default in Pytorch.
+    Adds the ability to pre-specify the weight matrix, use a non-trainable embedding and
+    catches some ill-advised use cases, such as 1 dimensional embeddings.
+
+    Parameters
+    ----------
+    num_embeddings:, int:
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: We should probably figure out markdown vs. rest sooner rather than later, so we can switch docstrings now, if we're going to switch.

If we're sticking with rest, this should be formatted like:
```
num_embeddings : ``int``
```
Note the space after the parameter name - it's important for the code that generates the docs.

File: allennlp/layers/embeddings.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            if self.padding_index is not None:
+                self.weight.data[self.padding_index].fill_(0)
+
+    def reset_parameters(self):
+        self.weight.data.normal_(0, 1)
+        if self.padding_index is not None:
+            self.weight.data[self.padding_index].fill_(0)
+
+    def forward(self, input):
+        padding_idx = self.padding_idx
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: `self.padding_index`

File: allennlp/layers/embeddings.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        self.num_embeddings = num_embeddings
+        self.embedding_dim = embedding_dim
+        self.padding_index = padding_index
+        self.max_norm = max_norm
+        self.norm_type = norm_type
+        self.scale_grad_by_freq = scale_grad_by_freq
+        self.sparse = sparse
+
+        if embedding_dim == 1:
+            raise ConfigurationError("There is no need to embed tokens if you "
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Why not?  I could be learning something like a tf-idf weight for each token.  Seems odd to put in this restriction, as it's unnecessary.

File: allennlp/layers/embeddings.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    An Embedding module.
+
+    """
+
+    def __init__(self,
+                 num_embeddings: int,
+                 embedding_dim: int,
+                 weight: torch.FloatTensor=None,
+                 padding_index: int=None,
+                 trainable: bool=True,
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: This field is missing from the docstring.

File: allennlp/layers/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        self.sparse = sparse
+
+        if embedding_dim == 1:
+            raise ConfigurationError("There is no need to embed tokens if you "
+                                     "are using a 1 dimensional embedding.")
+
+        if weight is None:
+            weight = torch.FloatTensor(num_embeddings, embedding_dim)
+            self.weight = torch.nn.Parameter(weight, requires_grad=trainable)
+            self.reset_parameters()
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: How does this work?  This is scary to me.

File: allennlp/layers/embeddings.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: I'd probably remove this line.

File: allennlp/layers/embeddings.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        else:
+            if weight.size() != (num_embeddings, embedding_dim):
+                raise ConfigurationError("A weight matrix was passed with contradictory embedding shapes.")
+            self.weight = torch.nn.Parameter(weight, requires_grad=trainable)
+            if self.padding_index is not None:
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Why do this?  Presumably you want a mask, right?  This seems like a pointless operation - it doesn't avoid errors with a softmax attention, for example.

File: allennlp/layers/embeddings.py (hunk 5 above)
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Oh, you defined this method below.  Ok, that makes more sense.  When I first saw this, I thought you were calling a super class method, and I was really nervous about using pytorch.

Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Do you even need this method?  Looks like it's just a single line, and you can move the padding fill to below the if/else block, because you do it in both.

Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: And if you do want it, I'd call it something like `initialize_parameters`.

File: allennlp/layers/embeddings.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                                   trainable: bool=True,
+                                   log_misses: bool=False):
+    """
+    Reads a pre-trained embedding file and generates a Keras Embedding layer that has weights
+    initialized to the pre-trained embeddings.  The Embedding layer can either be trainable or
+    not.
+
+    We use the DataIndexer to map from the word strings in the embeddings file to the indices
+    that we need, and to know which words from the embeddings file we can safely ignore.  If we
+    come across a word in DataIndexer that does not show up with the embeddings file, we give
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: ```s/DataIndexer/``Vocabulary``/ ```

File: allennlp/layers/embeddings.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    it a zero vector.
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Why a zero vector?  Why not a random vector?  (I know you're probably copying what we had before - that doesn't mean we should keep it that way.)

File: allennlp/layers/embeddings.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+                                   namespace: str="tokens",
+                                   trainable: bool=True,
+                                   log_misses: bool=False):
+    """
+    Reads a pre-trained embedding file and generates a Keras Embedding layer that has weights
+    initialized to the pre-trained embeddings.  The Embedding layer can either be trainable or
+    not.
+
+    We use the DataIndexer to map from the word strings in the embeddings file to the indices
+    that we need, and to know which words from the embeddings file we can safely ignore.  If we
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: We probably don't want to ignore words in the pretrained file that aren't in the vocab, but we can handle that later.

File: allennlp/layers/embeddings.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+                                       self.sparse)(input, self.weight)
+
+
+def get_pretrained_embedding_layer(embeddings_filename: str,
+                                   vocab: Vocabulary,
+                                   namespace: str="tokens",
+                                   trainable: bool=True,
+                                   log_misses: bool=False):
+    """
+    Reads a pre-trained embedding file and generates a Keras Embedding layer that has weights
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Keras?

File: allennlp/layers/embeddings.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    initialized to the pre-trained embeddings.  The Embedding layer can either be trainable or
+    not.
+
+    We use the DataIndexer to map from the word strings in the embeddings file to the indices
+    that we need, and to know which words from the embeddings file we can safely ignore.  If we
+    come across a word in DataIndexer that does not show up with the embeddings file, we give
+    it a zero vector.
+
+    The embeddings file is assumed to be gzipped, formatted as [word] [dim 1] [dim 2] ...
+    """
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Documenting parameters here would be good.

File: allennlp/layers/embeddings.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    words_to_keep = set(vocab.get_index_to_token_vocabulary(namespace).values())
+    vocab_size = vocab.get_vocab_size(namespace)
+    embeddings = {}
+    embedding_dim = None
+
+    # TODO(matt): make this a parameter
+    embedding_misses_filename = 'embedding_misses.txt'
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: We can probably just log misses using `logger.debug`, and remove the `log_misses` parameter entirely, and this variable.

File: allennlp/layers/embeddings.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    logger.info("Reading embeddings from file")
+    with gzip.open(embeddings_filename, 'rb') as embeddings_file:
+        for line in embeddings_file:
+            fields = line.decode('utf-8').strip().split(' ')
+            if embedding_dim is None:
+                embedding_dim = len(fields) - 1
+                assert embedding_dim > 1, "Found embedding size of 1; do you have a header?"
+            else:
+                if len(fields) - 1 != embedding_dim:
+                    # Sometimes there are funny unicode parsing problems that lead to different
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Can we do better here?  Maybe we can just import a library that reads glove vectors already, so we don't have to duplicate that functionality?

File: allennlp/layers/embeddings.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    # then filling in the word vectors we just read.
+    logger.info("Initializing pre-trained embedding layer")
+    if log_misses:
+        logger.info("Logging embedding misses to %s", embedding_misses_filename)
+        embedding_misses_file = codecs.open(embedding_misses_filename, 'w', 'utf-8')
+    embedding_matrix = torch.FloatTensor(vocab_size, embedding_dim).uniform_(-0.05, 0.05)
+
+    # Depending on whether the namespace has PAD and UNK tokens, we start at different indices,
+    # because you shouldn't have pretrained embeddings for PAD or UNK.
+    start_index = 0 if namespace.startswith("*") else 2
```
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: You should be able to always start at 0.  The check for `word in embeddings` will fail, and everything will be ok, because we just continue in that case.  And the namespace check is wrong here, anyways.

File: allennlp/layers/embeddings.py (hunk 3 above)
Comment (DeNeutoy) [inline allennlp/layers/embeddings.py]: Sure, I was just porting the previous embedding. I'll remove it.

File: allennlp/layers/embeddings.py (hunk 5 above)
Comment (DeNeutoy) [inline allennlp/layers/embeddings.py]: It looks like there is a technical reason to have this function if we ever need to call it to create parameters in the call of `forward`: https://discuss.pytorch.org/t/dynamic-parameter-declaration-in-forward-function/427/7

This isn't the reason it's used here though, but I was wondering why other pytorch layers use it and it looks like this might be why.

File: allennlp/layers/embeddings.py (hunk 7 above)
Comment (DeNeutoy) [inline allennlp/layers/embeddings.py]: Sure - do we want to do masking right from the off here? I can change this to exclusively providing a mask, but also I think it's better than not to have the padding index be all zeros. It immediately stands out if you are looking at a tensor, for example.

File: allennlp/layers/embeddings.py (hunk 5 above)
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: Interesting, ok.  But it's not a method defined on `Module`, so we can handle this however we want, right?

File: allennlp/layers/embeddings.py (hunk 9 above)
Comment (DeNeutoy) [inline allennlp/layers/embeddings.py]: Oh... That wasn't even true beforehand. Removed.

File: allennlp/layers/embeddings.py (hunk 7 above)
Comment (matt-gardner) [inline allennlp/layers/embeddings.py]: I'd worry about handling the masking later.

The only concern is that zeroing out the padding token might give you a false sense of security, but it's probably not a big deal.  Do what you think is best.

Comment (DeNeutoy) [inline allennlp/layers/embeddings.py]: Yes that is true, and it would make a few cases easy to miss. We can decide when we implement padding.

File: allennlp/layers/embeddings.py (hunk 5 above)
Comment (DeNeutoy) [inline allennlp/layers/embeddings.py]: Yes, I have removed it for now.
